### PR TITLE
73 fn:graphemes

### PR DESCRIPTION
**Pull request by Arithmeticus**
First draft of `fn:graphemes`, in response to discussion at #73 .

A battery of tests will be submitted as a PR to the qt4tests repository.

### CODE REVIEW

**Comment by Arithmeticus** [comment]
First draft of tests: https://github.com/qt4cg/qt4tests/pull/77

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 1]
```diff
@@ -28119,6 +28119,104 @@ return every($dl/*, fn($elem, $pos) {
...
+      <fos:rules>
+         <p>The <code>$options</code> argument, if present, defines additional parameters
+            controlling how the process is conducted. The <termref def="option-parameter-conventions"
+               >option parameter conventions</termref> apply.</p>
+         <p>If the one-argument version of the function is used, the result is the same as calling
+            the two-argument version, with the default <code>$options</code> settings. </p>
+         <p>The function returns a sequence of strings. Each string contains one or more <termref
+               def="character">character</termref>s that collectively constitute a single grapheme
+            cluster, as defined by <bibref ref="UNICODE-TR29"/>.</p>
+         <p>If <code>$value</code> is a zero-length string or the empty sequence, the function returns
```
Comment (ChristianGruen) [inline specifications/xpath-functions-40/src/function-catalog.xml:28152]: A minor issue: This condition could be limited to the empty sequence, as (I assume) the core split algorithm implies that a zero string will return zero characters.

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 2]
```diff
@@ -28119,6 +28119,104 @@ return every($dl/*, fn($elem, $pos) {
...
+            the empty sequence.</p>
+         
+         <fos:options>
+            <fos:option key="grapheme-clusters">
```
Comment (ChristianGruen) [inline specifications/xpath-functions-40/src/function-catalog.xml:28156]: Maybe just `clusters`.

I failed finding information on the two variants in the quoted reference. Could you possibly add some more details how a legacy implementations would differ from the default implementation?

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 3]
```diff
@@ -28119,6 +28119,104 @@ return every($dl/*, fn($elem, $pos) {
...
+                  </ulist>
+                  <p>Conforming implementations <rfc2119>may</rfc2119> support other grapheme
+                     cluster boundaries with implementation-defined semantics.</p>
+               </fos:meaning>
+               <fos:type>xs:string</fos:type>
+               <fos:default>extended</fos:default>
+            </fos:option>
+         </fos:options>
+      </fos:rules>
+      <fos:examples>
```
Comment (ChristianGruen) [inline specifications/xpath-functions-40/src/function-catalog.xml:28178]: Examples for both the Legacy and Extended variant would be helpful to understand the difference.

**Review by ChristianGruen — CHANGES_REQUESTED** [review]

**Comment by michaelhkay** [comment]
It might be helpful in the examples to indicate what the actual codepoints in the input strings are, rather than relying on the way the browser renders them. For example, some of the characters might be in composed or decomposed form, and we can't tell just by looking at them. One solution might be to write the example as `let $s := "..." return [string-to-codepoints($s), graphemes($s)]`.

File: specifications/xpath-functions-40/src/function-catalog.xml (hunk 2 above)
Comment (Arithmeticus) [reply specifications/xpath-functions-40/src/function-catalog.xml:28156]: see question raised below

File: specifications/xpath-functions-40/src/function-catalog.xml (hunk 3 above)
Comment (Arithmeticus) [reply specifications/xpath-functions-40/src/function-catalog.xml:28178]: see question raised below

**Comment by Arithmeticus** [comment]
Regarding legacy versus extended grapheme clusters: I would like to suggest that we can drop the legacy as a required option.

> The Unicode Standard provides a default algorithm for determining grapheme cluster boundaries; the default grapheme clusters are also known as **extended grapheme clusters**. For backwards compatibility with earlier versions of this specification, the Standard also defines a profile for **legacy grapheme clusters**.

Of course, nothing forbids an implementer from supporting legacy, per the options map rules.

That will simplify the specs and tests, and not saddle implementations with legacy support.  Any objections?

File: specifications/xpath-functions-40/src/function-catalog.xml (hunk 1 above)
Comment (Arithmeticus) [reply specifications/xpath-functions-40/src/function-catalog.xml:28152]: Here I tried merely to follow `fn:characters`. If `fn:graphemes('')` should entail `''` then `fn:characters` should do so as well, no?

**Comment by ChristianGruen** [comment]
I suggest dropping the options argument. The specification should ideally not motivate implementors to do things differently unless there are compelling reasons to do so.

File: specifications/xpath-functions-40/src/function-catalog.xml (hunk 1 above)
Comment (ChristianGruen) [reply specifications/xpath-functions-40/src/function-catalog.xml:28152]: I agree. Consistency wins.

**Review by ChristianGruen — APPROVED** [review]

**Comment by ndw** [comment]
The CG agreed to merge this PR at [meeting 077](https://qt4cg.org/meeting/minutes/2024/05-14.html)

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 4]
```diff
@@ -28168,6 +28168,92 @@ return every($dl/*, fn($elem, $pos) {
...
+               <fos:result>$crlf</fos:result>
+               <fos:postamble>Carriage return + line feed, two characters, one
+                  grapheme</fos:postamble>
+            </fos:test>
+         </fos:example>
+         <fos:example>
+            <fos:test>
+               <fos:expression>graphemes(char(0x1F476) || char(0x200D) 
+|| char(0x1F6D1))</fos:expression>
+               <fos:result>char(0x1F476) || char(0x308) || char(0x1F6D1)</fos:result>
```
Comment (gimsieke) [inline specifications/xpath-functions-40/src/function-catalog.xml:28232]: Shouldn’t this `char(0x308)` read `char(0x200D)`?

**Review by gimsieke — COMMENTED** [review]
Since this has already been merged, I should probably issue a new PR?

**Comment by ChristianGruen** [comment]
> Since this has already been merged, I should probably issue a new PR?

Yes, that’ll probably be the easiest way.

**Pull request merged**